### ISSUE TITLE
Module 2 Feedback 

### ISSUE THREAD

**Issue by hfboyce** [UBC-MDS/introduction-machine-learning#19]
Please see the deployed link [here](https://intro-machine-learning.netlify.app/en/module2).
@kvarada it's important that we are both seeing the same version as sometimes your browser will not update and it can be very frustrating. Make sure to clear to cookies for this site from now on. 
This module should have 21 "exercises" (this is what we call each of these numbered sections). 

The latest change was in Exercises 21 the learning outcomes should NOT include the bullet: 
**Split a dataset into train and test sets using train_test_split() function.**

Please feel free to point out everything you can so we can improve this and produce the best product. 
Thanks :)

**Comment by mgelbart** [comment]
@kvarada I'll leave this to you, but let me know if there's anything you want me to look at.

**Comment by kvarada** [comment]
Sounds good. I'll get to this in the afternoon.

**Comment by kvarada** [comment]
Great job with the module @hfboyce ! I see a bit of "Hayley touch" :).  

Here are my comments for improvement for [module2](https://intro-machine-learning.netlify.app/en/module2)

#### 0 
- [ ] 0.2 Explain the different between --> Explain the difference between  
- [ ] 0.3 Let's Start! --> Fine but not sure if we need capitalization for "Start" as it's a sentence and not a title. 

#### 1
- [ ] 1.2 Not sure why you have "recap" here. Probably you used the same picture in module 1. 
- [ ] 1.3 Not sure whether we want to get into how many combinations we can get here for these students. @mgelbart what do you think? 
- [ ] 1.3 After saying that rule-based approach get unwieldy, I would say that decision tree gives you a better way to do this before you go into binarizing the data and that's what we are going to talk about next.   
- [ ] 1.4 Where is the code to open the toy data CSV? Is it in the previous module? Probably we should open it again here and display the shape of the CSV?
- 1.6 Beautiful picture of trees :). I would like to go there!
- 1.7 and 1.8 Nice pictures! 
- [ ] 1.8 Decision trees Terminology --> Decision trees terminology (I think we are not capitalizing titles elsewhere.)

#### 2
- [ ] I suggest changing both questions. Although `sklearn` doesn't support categorical variables in decision trees, decision trees are capable of handling them. For example you could have a categorical variable "weather" with three possible values (e.g., warm, hot, and cold) and then there would be 3 children to a node and it would represent if/elif/else statement. What do you think @mgelbart?  

#### 3
- Good to have a question on depth of the tree, as they will be working with `max_depth` later in the module. 

#### 4
- [ ] 4.5 I suggest we use Mike's `display_tree` function to get rid of all the unnecessary stuff from `graphviz` trees. You can find it [here](https://github.com/UBC-CS/cpsc330/blob/master/lectures/02_decision-trees.ipynb)
- [ ] 4.8 Love the animation! Might be nice to show the full path till the leaf node for the left branch of the root node as well? Or at least show dots to indicate that we are not showing a complete tree here.  

#### 5 
- [ ] Typos: Berry = 1 --> Berry == 1 and green = 1 -- >  green == 1
- Like the questions otherwise. 

#### 6
- [ ] 1. I would be specific here and say `DummyClassifier` instead of baseline models. 
- [ ] 2. I would use `fit` and `predict` here. So something like: We need to `fit` our decision tree model before calling `predict`.  
- [ ] 3. I selected False and it said. "Incorrect. Great!" :). This is a tricky question. I assume you mean more than one predictions for the same example. The problem is that if there is a tie between attributes when fitting this may happen. I would change this question. Probably I would ask something about whether they would always give you the correct answer or not. 

#### 7
- Nice exercise! 
- [ ] Fit your model on the objects X and y and then predict on the target column y. --> predict on X? 
- [ ] Tell them to store the predictions in a variable called `predicted`. 

#### 8 
- [ ] 8.1 Just making sure. Have we introduced the term "features" before?
- [ ] 8.3 See my comment 4.5
- 8.8 Thanks for `altair` plots!! 

#### 9 
- [ ] 9.1 Are we explicitly saying what a decision stump is in 8? I guess it'll be in the transcript. 
- [ ] 9.2 I would be consistent with the capitalization in the feature names (e.g., in the tree it's "Water Content" and in the question it's "water content"). 
- [ ] 9.3. Probably, I would say "On what value of the above feature is the decision boundary splitting?" 

#### 10
- Nice example! 
- [ ] We are trying to predict a player’s in this problem: -- > We are trying to predict a player’s XXX (position/role?) in this problem:
- [ ] 10.3 It's tricky here because you starting the y axis with 170 and it's not labeled. 

#### 11 
- [ ] We are trying to predict a player’s in this problem: -- > We are trying to predict a player’s XXX (position/role?) in this problem:

#### 14
- [ ] Question 1 and 2: When I try to answer Question 2, it changes my answer to Question 1. 

#### 15
- [ ] I faced a few problems in this exercise. I got connection lost error once. Also, I don't know whether I should be displaying `tree_score` or not. Also, do we need to call predict here if we are doing `score`? It could be my English problem but I find "and then predict on the target column y" a bit confusing. For me, we are predicting on `X` to get hat{y}
- [ ] Question 1: Will increasing the value of max_depth make the accuracy of the model increase or decrease? --> (suggestion) Will increasing the value of max_depth increase or decrease the accuracy of the model?
- [ ] Question 2: same suggestion as of Q1

#### 18
- [ ] In the coding question it was not clear whether I got the right answer or not. 

#### 20
- [ ] 20.1 No, it doesn't generalize well to other unseen data --> (suggestion) No, it may not generalize well to other unseen data. 

Great work overall!

**Comment by kvarada** [comment]
Btw, it took me ~2 hours. Probably because it was my first time (and I am also a bit slow in general :)).

**Comment by mgelbart** [comment]
> Btw, it took me ~2 hours. Probably because it was my first time (and I am also a bit slow in general :)).

Yeah I wasn't tracking my hours back then but I think it probably took me ~2 hours as well at the beginning, but then I got faster. 

>  I suggest changing both questions. Although sklearn doesn't support categorical variables in decision trees, decision trees are capable of handling them. For example you could have a categorical variable "weather" with three possible values (e.g., warm, hot, and cold) and then there would be 3 children to a node and it would represent if/elif/else statement. What do you think @mgelbart?

My take is that you'd always have 2 children but you'd have split rules like "weather == warm?" rather than having to one-hot encode weather, which is a limitation of scikit-learn. I guess there's this new CatBoost which handles categorical variables directly with some sort of gradient boosted trees. So personally I'm OK with the second question here. For the first question, it's a bit unclear to me whether children means direct children or all descendants. I agree with @kvarada on removing the first one and I'm fine either keeping or deleting the second one.

**Comment by kvarada** [comment]
Oh I see what you mean. I was thinking about pictures such as this, which are not uncommon to represent decision trees.  

<img width="597" alt="Screen Shot 2020-10-08 at 3 14 57 PM" src="https://user-images.githubusercontent.com/22737743/95518991-143a7400-0979-11eb-9d8b-5ccc200c8b70.png">

[Source](http://www2.cs.uh.edu/~ceick/ai/dectree.pdf)

**Comment by mgelbart** [comment]
Yeah makes sense. I guess under the hood they would probably be two separate splits, one for == Sunny and then, on the False branch, another for == Overcast. But I agree it's getting into technicalities which are besides the point of what we're teaching here.

**Comment by hfboyce** [comment]
So take both out or just the first one? I'll can replace them with something else.

**Comment by mgelbart** [comment]
I think the second one is keep-able if we remove the 2nd option (if/elif/else), to reduce confusion.

**Comment by hfboyce** [comment]
- plot classifier lims 
- change code in display tree.

**Comment by kvarada** [comment]
@hfboyce I used Module 2 exercises in my class activities today. Seems like the students enjoyed them :tada:. A couple of things came up: 

- Many of them were not able to run code exercises; they were getting binder hub error. 
- The autograding for 7.3 is probably wrong?
- In 11 we are showing them two trees: one with class in each node and other without class. It was confusing for the students.

**Comment by hfboyce** [comment]
@kvarada 

> Many of them were not able to run code exercises; they were getting binder hub error.

This may have something to do either with their browser, network or where they are located geographically. Do you know what browser they were using or where they are living? 

> The autograding for 7.3 is probably wrong?

Yes Elijah has yet to make the tests for Module 2 so this autograding is 100% wrong. 

> In 11 we are showing them two trees: one with class in each node and other without class. It was confusing for the students.

I've updated this! Thank you!
